src/bun.js/bindings/JSBuffer.cpp

```diff
@@ -1350,108 +1403,97 @@ static int64_t lastIndexOf(const uint8_t* thisPtr, int64_t thisLength, const uin
-            } else if (byteOffset_ < 0) {
-                byteOffset = length + static_cast<int64_t>(byteOffset_);
-            } else {
-                byteOffset = static_cast<int64_t>(byteOffset_);
-            }
-
-            if (last) {
-                if (byteOffset < 0) {
-                    return -1;
-                } else if (byteOffset > length - 1) {
-                    byteOffset = length - 1;
-                }
-            } else {
-                if (byteOffset <= 0) {
-                    byteOffset = 0;
-                } else if (byteOffset > length - 1) {
-                    return -1;
-                }
-            }
+    if (std::isnan(byteOffsetD)) byteOffsetD = dir ? 0 : byteLength;
 
-            if (callFrame->argumentCount() > 2) {
-                EnsureStillAliveScope encodingValue = callFrame->uncheckedArgument(2);
-                if (!encodingValue.value().isUndefined()) {
-                    encoding = parseEncoding(lexicalGlobalObject, scope, encodingValue.value());
-                    RETURN_IF_EXCEPTION(scope, -1);
-                }
+    if (valueValue.isNumber()) {
+        ssize_t byteOffset = indexOfOffset(byteLength, byteOffsetD, 1, dir);
+        if (byteOffset == -1) return -1;
+        uint8_t byteValue = (valueValue.toInt32(lexicalGlobalObject)) % 256;
+        RETURN_IF_EXCEPTION(scope, -1);
+        if (last) {
+            for (int64_t i = byteOffset; i >= 0; --i) {
+                if (byteValue == typedVector[i]) return i;
             }
+        } else {
+            const void* offset = memchr(reinterpret_cast<const void*>(typedVector + byteOffset), byteValue, byteLength - byteOffset);
+            if (offset != NULL) return static_cast<const uint8_t*>(offset) - typedVector;
         }
+        return -1;
     }
 
-    if (value.isString()) {
-        auto* str = value.toStringOrNull(lexicalGlobalObject);
-        RETURN_IF_EXCEPTION(scope, -1);
+    WTF::String encodingString;
+    if (!encodingValue.isUndefined()) {
+        encodingString = encodingValue.toWTFString(lexicalGlobalObject);
+        RETURN_IF_EXCEPTION(scope, {});
+        encoding = parseEnumeration2(*lexicalGlobalObject, encodingString);
+    } else {
+        encoding = BufferEncodingType::utf8;
+    }
 
-        JSC::EncodedJSValue encodedBuffer = constructFromEncoding(lexicalGlobalObject, str, encoding);
+    if (valueValue.isString()) {
+        if (!encoding.has_value()) {
+            return Bun::ERR::UNKNOWN_ENCODING(scope, lexicalGlobalObject, encodingString);
+        }
+        auto* str = valueValue.toStringOrNull(lexicalGlobalObject);
+        RETURN_IF_EXCEPTION(scope, -1);
+        ssize_t byteOffset = indexOfOffset(byteLength, byteOffsetD, str->length(), dir);
+        if (byteOffset == -1) return -1;
+        if (str->length() == 0) return byteOffset;
+        JSC::EncodedJSValue encodedBuffer = constructFromEncoding(lexicalGlobalObject, str, encoding.value());
```

Comment:
why does this return an EncodedJSValue?

Reply:
it calls 4 separate functions based on the encoding and two are in zig it seems